### PR TITLE
Add node type support for GraphQL

### PR DESCRIPTION
**Pull request by Jotschi**
## Abstract

* Adds nodeType argument at various places within the GraphQL API in order to select the type of the content (draft/published)
* Fixes various permission issues with READ and READ_PUBLISH permissions

### General

* [x] Added abstract that describes the change
* [x] Added changelog entry to `/CHANGELOG.adoc`
* [x] Ensured that the change is covered by tests
* [x] Ensured that the change is documented in the docs

### On API Changes

* [x] Checked if the changes are breaking or not
* [x] Added GraphQL API if applicable
* [x] Added Elasticsearch mapping if applicable

### CODE REVIEW

**Comment by Jotschi** [comment]
Relates to https://github.com/gentics/mesh-incubator/issues/166

**Comment by Jotschi** [comment]
Testcases:

```
{
  me {
    # Via Node Reference
    nodeReference {
      uuid
    }
  }
  schema(name: "vehicle") {
    uuid
    # Via schema nodes
    nodes {elements {uuid}}
  }
  
  # Via tag
  tag(name: "Kerosene") {
    name
    nodes {
      elements {uuid}
    }
  }
  # Via Tags
  tags {
    elements {
      name
      nodes {
        elements {
          uuid
        }
      }
    }
  }
  # Via UUID
  node(uuid: "ca6c7df3f45b48d4ac7df3f45ba8d42f") {
    uuid
    path
    # Nested via Node
    node (lang: "ru") {
      uuid
    }
    
    ... on vehicle {
    	fields {
        # Via Node Reference
      	vehicleImage {
          uuid
        }
    	}
    }
  }
  # Via Path
  p2: node(path: "/automobiles")  {
    uuid
  }
  # Via Nodes
  nodes {
    elements {
      uuid
    }
  }
  # Via ES query
  n2:nodes(query:"") {
    elements {
      uuid
    }
  }
}
```

* Via `.me.nodeReference` (user node reference)
* Via `.tag.nodes` (tagged nodes)
* Via `.schema.nodes` (schema nodes)
* Via `.node` (UUID)
* Via `.node` (Path)
* Via `.nodes` (ES query)
* Via `.nodes` Collection

**Comment by Jotschi** [comment]
TODO:
* Add type handling to NodeSearchHandler
* Check whether micronode type handling needs custom code

File: elasticsearch/src/main/java/com/gentics/mesh/search/index/node/NodeSearchHandler.java [hunk 1]
```diff
@@ -83,6 +84,7 @@ public NodeSearchHandler(SearchProvider searchProvider, Database db, NodeIndexHa
 		if (log.isDebugEnabled()) {
 			log.debug("Invoking search with query {" + query + "} for {Containers}");
 		}
+		// TODO !! pass along type parameter to reduce / correct selected indices
```
Comment (philippguertler) [inline elasticsearch/src/main/java/com/gentics/mesh/search/index/node/NodeSearchHandler.java:87]: Override getIndexHandler in NodeSearcHandler. Implement `getIndicesForSearch` with additional type as parameter

File: verticles/graphql/src/main/java/com/gentics/mesh/graphql/model/NodeReferenceIn.java [hunk 2]
```diff
@@ -33,25 +32,30 @@ private NodeReferenceIn(NodeContent node, NodeGraphField nodeGraphField) {
...
-				// TODO extract this common behaviour, see https://github.com/gentics/mesh/issues/698
-				return version.equals("draft") && container.isType(DRAFT, branchUuid) && user.hasPermission(node, GraphPermission.READ_PERM) ||
-					version.equals("published") && container.isType(PUBLISHED, branchUuid) && user.hasPermission(node, GraphPermission.READ_PUBLISHED_PERM);
-			}).findAny())
-			.map(referencingContent -> new NodeReferenceIn(
-				new NodeContent(null, referencingContent, content.getLanguageFallback()),
-				ref
-			)));
+				.filter(container -> {
+					if (type == DRAFT && container.isDraft()) {
```
Comment (philippguertler) [inline verticles/graphql/src/main/java/com/gentics/mesh/graphql/model/NodeReferenceIn.java:46]: Is this branch aware?

File: core/src/test/java/com/gentics/mesh/core/graphql/GraphQLSearchPermissionTest.java [hunk 3]
```diff
@@ -47,27 +54,35 @@ public void testReadPublishPerm() throws Exception {
...
-		}
-		GraphQLResponse response = call(() -> client().graphqlQuery(PROJECT_NAME, getGraphQLQuery(queryName), params));
+	private void assertSearch(String queryName, int expectedDraftResults, int expectedPublishedResults) throws IOException {
+		GraphQLResponse response = call(() -> client().graphqlQuery(PROJECT_NAME, getGraphQLQuery(queryName)));
 		JsonObject json = new JsonObject(response.toJson());
-		System.out.println(json.encodePrettily());
 		assertThat(json).compliesToAssertions(queryName);
-		assertThat(((JsonArray)JsonPointer.from("/data/result/elements").queryJson(json)).size()).isEqualTo(expectedResults);
+
+		JsonArray draftElements = (JsonArray) JsonPointer.from("/data/dresult/elements").queryJson(json);
```
Comment (philippguertler) [inline core/src/test/java/com/gentics/mesh/core/graphql/GraphQLSearchPermissionTest.java:73]: Also write test which chooses version with query parameter

File: core/src/test/resources/graphql/node-elasticsearch-perm-query [hunk 4]
```diff
@@ -1,15 +1,37 @@
...
+			node {
+				uuid
+			}
 			language
+			version
 		}
 	}
 
-	result: nodes(query: "{ \"query\": { \"match\": { \"fields.slug\": \"blaar\" } } }") {
+	presult: nodes(nodeType: published, query: "{ \"query\": { \"match\": { \"fields.slug\": \"blaar\" } } }") {
```
Comment (philippguertler) [inline core/src/test/resources/graphql/node-elasticsearch-perm-query:13]: Rename nodeType to version

File: core/src/test/java/com/gentics/mesh/core/graphql/GraphQLNodePermissionTest.java [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+			RolePermissionRequest permRequest2 = new RolePermissionRequest();
+			permRequest2.setRecursive(false);
+			permRequest2.getPermissions().set(Permission.READ, true);
+			permRequest2.getPermissions().setOthers(false);
+			adminCall(() -> client().updateRolePermissions(roleUuid(),
+				"/projects/" + PROJECT_NAME + "/nodes/" + NODE_1B_UUID, permRequest2));
+		}
+	}
+
+	public static enum TestField {
```
Comment (philippguertler) [inline core/src/test/java/com/gentics/mesh/core/graphql/GraphQLNodePermissionTest.java:172]: Rename to TestQueries

**Pull request merged**